### ISSUE TITLE
Feature request: use the Artifacts system to provide a C compiler on Linux and macOS

### ISSUE THREAD

**Issue by DilumAluthge** [JuliaLang/PackageCompiler.jl#316]
On Windows, PackageCompiler now uses the Artifacts system to provide a C compiler.

Would it be possible to use the Artifacts system to also provide a C compiler on Linux and macOS?

That would make it possible to build custom Julia sysimages on any of the three major operating systems without needing any additional dependencies, which I think would be really cool!

**Comment by KristofferC** [comment]
It is so super easy to get a compiler there so bit sure the extra download is worth it. For people making containers it might make them significantly larger?

**Comment by DilumAluthge** [comment]
That's a good point. It would be nice to have as an optional feature maybe?

Is there a way to download Artifacts on-demand? As in, by default we don't download a C compiler on Linux or macOS, but if the user runs `PackageCompiler.download_compiler()`, then it downloads the artifact.

**Comment by KristofferC** [comment]
Yes, if the artifact is `lazy=true´ (this is actually the case already on Windows) so if you have `JULIA_CC` set it just uses that one (not documented right now though).

**Comment by DilumAluthge** [comment]
> Yes, if the artifact is `lazy=true` in the Artifact file (this is actually the case already on Windows) so if you have `JULIA_CC` set it just uses that one (not documented right now though).

So perhaps we could add compilers for Linux and macOS to the Artifact file with `lazy=true`. And then we would have an exported function `PackageCompiler.download_compiler()` that calls the `Pkg.Artifacts.artifact"..."` macro and sets `ENV["JULIA_CC"]`.

Then the workflow (for people that want to download the compiler on Linux and macOS) would be:
```julia
julia> using PackageCompiler

julia> download_compiler()

julia> create_sysimage(...)
```

**Comment by DilumAluthge** [comment]
In addition to setting `ENV["JULIA_CC"]`, we'd probably need to set `C_INCLUDE_PATH` to make sure that the `gcc` we provide via Artifacts is able to find the correct header files.

**Comment by DilumAluthge** [comment]
I tried to get this working by providing Clang binaries (#326), but I kept running into strange issues, such as Clang trying (and failing) to find ncurses.
